iceoryx_hoofs/test/moduletests/test_cxx_expected.cpp

```diff
@@ -187,58 +202,64 @@ TEST_F(expected_test, ConstCreateLValueAndGetErrorResultsInCorrectError)
     EXPECT_THAT(sut.get_error().m_b, Eq(232));
 }
 
-TEST_F(expected_test, CreateWithValueAndMoveCtorLeadsToInvalidState)
+TEST_F(expected_test, CreateWithValueAndMoveCtorLeadsToMovedSource)
 {
-    auto sut = expected<int, TestClass>::create_value(177);
-    auto movedValue{std::move(sut)};
-    IOX_DISCARD_RESULT(movedValue);
-    ASSERT_TRUE(sut.has_error());
-    EXPECT_THAT(sut.get_error(), Eq(TestClassAdapter::getInvalidState()));
-}
+    constexpr int A{177};
+    constexpr int B{188};
+    auto sutSource = expected<NonTrivialTestClass, int>::create_value(A, B);
+    auto sutDestination{std::move(sutSource)};
 
-TEST_F(expected_test, CreateWithErrorAndMoveCtorLeadsToInvalidState)
-{
-    auto sut = expected<int, TestClass>::create_error(22, 33);
-    auto movedValue{std::move(sut)};
-    IOX_DISCARD_RESULT(movedValue);
-    ASSERT_TRUE(sut.has_error());
-    EXPECT_THAT(sut.get_error(), Eq(TestClassAdapter::getInvalidState()));
+    ASSERT_FALSE(sutSource.has_error());
+    EXPECT_TRUE(sutSource.value().m_moved);
+    ASSERT_TRUE(!sutDestination.has_error());
+    EXPECT_FALSE(sutDestination.value().m_moved);
+    EXPECT_EQ(sutDestination.value().m_a, A);
+    EXPECT_EQ(sutDestination.value().m_b, B);
 }
 
-TEST_F(expected_test, CreateWithValueAndMoveAssignmentLeadsToInvalidState)
+TEST_F(expected_test, CreateWithErrorAndMoveCtorLeadsToMovedSource)
 {
-    auto sut = expected<int, TestClass>::create_value(73);
-    auto movedValue = std::move(sut);
-    IOX_DISCARD_RESULT(movedValue);
-    ASSERT_TRUE(sut.has_error());
-    EXPECT_THAT(sut.get_error(), Eq(TestClassAdapter::getInvalidState()));
-}
+    constexpr int A{22};
+    constexpr int B{33};
+    auto sutSource = expected<int, NonTrivialTestClass>::create_error(A, B);
+    auto sutDestination{std::move(sutSource)};
 
-TEST_F(expected_test, CreateWithErrorAndMoveAssignmentLeadsToInvalidState)
-{
-    auto sut = expected<int, TestClass>::create_error(44, 55);
-    auto movedValue = std::move(sut);
-    IOX_DISCARD_RESULT(movedValue);
-    ASSERT_TRUE(sut.has_error());
-    EXPECT_THAT(sut.get_error(), Eq(TestClassAdapter::getInvalidState()));
+    ASSERT_TRUE(sutSource.has_error());
+    EXPECT_TRUE(sutSource.get_error().m_moved);
```

Comment:
I don't like how the `expected` usage is showcased here. We typically use an `enum class` as `ErrorType`. An `enum class` can't have a custom move c'tor, right? How can one find out if a moved `expected` is still valid to use?  What we want is defined behaviour aka destructive move, I don't yet see how we can achieve this with the proposed changes in this PR. Please add a test case which uses an `enum class` as `ErrorType`.

Reply:
Destructive move is a language feature and not a library feature and C++ will (most probably) never have destructive moves. Destructive moves means that the source is not accessible anymore after the move but with C++ that is not the case. All we can do is place moved from objects in a `valid but unspecified state`. In this case, it contains a moved inner type. This is not only valid but also a specified state of the `expected` and therefore there is no undefined behavior. The moved from inner type must of course also obey the rule to be placed in a `valid but unspecified state`. In our case we would of course specify the state since we do not want undefined behavior.

This is a test to check that the `expected` indeed moves the inner type instead of e.g. just doing a copy. Having a move test where a `enum class` is used as `ErrorType` has no value since PODs are copied on move and therefore the moved from value will remain the same and we cannot check if a move occurred.

A moved expected behaves exactly the same as e.g.
```cpp
Foo foo;
fooExp.and_then([&] (auto& f) { foo = std::move(f); });
```
or
```cpp
auto foo = std::move(fooExp.value());
```
or
```cpp
Bar make_bar(Foo& f) {
    return Bar(std::move(f));
}

auto b = make_bar(fooExp.value());
```
or
```cpp
auto err = std::move(fooExp.get_error());
```

In the end the `expected` is more of a transport type than a container type. It's not stored in classes and used as return value. The usage is therefore local an easy to reason about. No need to artificially make it more complex when there are still many ways which lead to an `expected` with a moved inner type. In the end, I didn't even had to change anything in the code while removing this requirement, except the `INVALID_STATE` value, which shows that this is not handled in the code anyway because we do not access moved from objects and also check this in our tests. Furthermore, it behaves exactly the same as a moved `variant`.

Reply:
> Destructive move is a language feature and not a library feature and C++ will (most probably) never have destructive moves.

You're right. What I meant was a "invalidating move". We already do this throughout the code base and it's important e.g. when classes handle resources. 

> The moved from inner type must of course also obey the rule to be placed in a `valid but unspecified state`

My philosophy so far is, that with iceoryx we go one step further and for non-POD types go into a "invalid state and specified state".

> How can one find out if a moved `expected` is still valid to use?

What I realised with the `enum class`'s is, that it's not actually a dangerous thing because if a user accesses a moved-from `expected` which contains an error, he can only mistake the error and not by mistake take the happy path. 

>  I didn't even had to change anything in the code while removing this requirement

That's a good finding and very smelly. Thanks for taking the time to remove the `INVALID_STATE`!